src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java

```diff
@@ -74,16 +74,30 @@ public boolean start(ClickHouseSinkConfig csc) {
 
         LOGGER.debug("Ping was successful.");
 
-        List<Table> tableList = chc.extractTablesMapping();
-        if (tableList.isEmpty()) {
+        this.updateMapping();
+        if (mapping.isEmpty()) {
             LOGGER.error("Did not find any tables in destination Please create before running.");
             return false;
         }
 
-        for (Table table: tableList) {//TODO: Should we pull ALL tables in memory?
+        return true;
+    }
+    public void updateMapping() {
+        LOGGER.debug("Update table mapping.");
+
+        // Getting tables from ClickHouse
+        List<Table> tableList = this.chc.extractTablesMapping();
+        if (tableList.isEmpty()) {
+            return;
+        }
+
+        // Adding new tables to mapping
+        // TODO: check Kafka Connect's topics name or topics regex config and
+        // only add tables to in-memory mapping that matches the topics we consume.
+        // TODO: remove dropped tables from mapping
+        for (Table table: tableList) {
             this.mapping.put(table.getName(), table);
```

Comment:
This logic only adds new tables but does not remove the old ones (we are not clearing the old mapping potential memory leak)

Reply:
Made mapping thread-safe and added code to remove dropped tables. Please take a look.

Reply:
@hekike I would take another direction here keep this.mapping = new HashMap<>();
In the updateMapping create a new mapping and only in the end of the method switch between the references 
Read https://stackoverflow.com/questions/44283378/why-is-reference-assignment-atomic-in-java